Tools/clinic/clinic.py

```diff
@@ -160,25 +180,21 @@ def warn_or_fail(
     line_number: int | None = None,
 ) -> None:
     joined = " ".join([str(a) for a in args])
-    add, output = text_accumulator()
-    if fail:
-        add("Error")
-    else:
-        add("Warning")
     if clinic:
         if filename is None:
             filename = clinic.filename
         if getattr(clinic, 'block_parser', None) and (line_number is None):
             line_number = clinic.block_parser.line_number
-    if filename is not None:
-        add(' in file "' + filename + '"')
-    if line_number is not None:
-        add(" on line " + str(line_number))
-    add(':\n')
-    add(joined)
-    print(output())
     if fail:
-        sys.exit(-1)
+        raise ClinicError(joined, lineno=line_number, filename=filename)
+    else:
+        msg = "Warning"
+        if filename is not None:
+            msg += f" in file {filename!r}"
+        if line_number is not None:
+            msg += f" on line {line_number}"
+        msg += f": {joined}"
+        warnings.warn(msg, ClinicWarning)
```

Comment:
I'm wondering if using `warnings.warn()` is maybe a little overengineered for our use case. Does it get us anything over just printing to stdout?

Also: we appear to only have two uses of `warn()` inside `clinic.py`, and I don't think either of them is explicitly tested at the moment; we should probably merge #107514 first, which explicitly tests one of them

Reply:
Yes, let's add tests first.

Reply:
I'm fine with using stdout. This is not a library, it's just a tool.

Reply:
Perhaps stderr is better.